flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java

```diff
@@ -106,8 +240,191 @@
 		return deserializationSchema;
 	}
 
-	@Override
-	public String explainSource() {
-		return "";
+	/**
+	 * Assigns ingestion time timestamps and watermarks.
+	 */
+	public static class IngestionTimeWatermarkAssigner implements AssignerWithPeriodicWatermarks<Row> {
+
+		private long curTime = Long.MIN_VALUE;
+
+		@Override
+		public long extractTimestamp(Row element, long previousElementTimestamp) {
+			long t = System.currentTimeMillis();
+			if (t > curTime) {
+				curTime = t;
+			}
+			return curTime;
+		}
+
+		@Nullable
+		@Override
+		public Watermark getCurrentWatermark() {
+			return new Watermark(curTime - 1);
+		}
+	}
+
+	protected AssignerWithPeriodicWatermarks<Row> getAssigner() {
+		return this.timestampAssigner;
+	}
+
+	/**
+	 * Checks that the provided row time attribute is valid, determines its position in the schema,
+	 * and adjusts the return type.
+	 *
+	 * @param rowtime The attribute to check.
+	 */
+	private void configureRowTimeAttribute(String rowtime) {
+		Preconditions.checkNotNull(rowtime, "Row time attribute must not be null.");
+
+		if (this.ingestionTimeAttribute != null) {
+			throw new ValidationException(
+				"You can only specify a row time attribute OR an ingestion time attribute.");
+		}
+
+		if (this.rowTimeAttribute != null) {
+			throw new ValidationException(
+				"Row time attribute can only be specified once.");
+		}
+
+		// get current fields
+		String[] fieldNames = ((RowTypeInfo) this.getReturnType()).getFieldNames();
+		TypeInformation[] fieldTypes = ((RowTypeInfo) this.getReturnType()).getFieldTypes();
+
+		// check if the rowtime field exists and remember position
+		this.rowtimeFieldPos = -1;
```

Comment:
I'm wondering why we need to remove the field here and add it back later on. Changing the orders of the fields seems problematic and can potentially break serialization (in very hacky cases).

Another question is that to which extent a customized timestamp assigner can reuse the code here? Is it possible to implement it as a decorator of the table source? That way it opens up the possibilities to reuse the code for other table sources.

Reply:
Yes, it's not very nice to move the time attribute into the `StreamRecord`, remove the `Row` field, and copy the `StreamRecord` timestamp back into the `Row`. That's just how the current interface is designed. However, this will be changed with [FLINK-7446](https://issues.apache.org/jira/browse/FLINK-7446). 

The issue of watermark generation (timestamps don't need to be generated as they are already expected to be in the `Row`) will be addressed by [FLINK-7548](https://issues.apache.org/jira/browse/FLINK-7548). 

We have to see how much of FLINK-7446 and FLINK-7548 we can solve before Flink 1.4.0. I did this PR to have at least some time attribute support for KafkaTableSources in 1.4.0. Do you need more specialized watermark generators for your use case?

Reply:
What about accelerating the efforts on FLINK-7446 and FLINK-7548?

Unfortunately our use cases are a little bit more than that thus this PR will not solve the problem out of the box. In one use case we have a timestamp that is a `double` instead of a `bigint`, in another use case we have the timestamp sit in a nested structure. That's the reason why I'm more inclined for a decorator-based approach which is more easy to customize.

Reply:
I think you are right. We should not ask each `TableSource` to assign watermarks. 

A decorator approach, i.e., a watermark generation wrapper around a `TableSource` would be one option. Alternatively, we could also give the responsibility for watermark generation to the scan operator. All the `TableSource` would need to provide is the name of the field for which watermarks are generated and the watermark strategy ascending, bounded, or custom). The generation would be handled completely internally. At that point, we can also access all fields of any type (Row, Pojo, Tuple, ...), nested or flat, via code generation. So, we would need to extend or replace the `DefinedRowtimeAttribute` interface because we need information about the watermark strategy (ascending, bounded, or custom). What do you think @haohui?

I already started to work on FLINK-7446 but will also not too much time in the next weeks due to conferences. I hope that we can address these issues with the next release.

Reply:
Currently in our internal system we worked around this problem by decorating the data stream return by Kafka. Having the scan operator to assign the watermark seems pretty neat. +1 for that.

It might be cleaner to put both the assignment of the timestamp (i.e., `TimestampAssigner`) and the naming of the timestamp (i.e., `DefinedRowtimeAttribute` and `DefinedProctimeAttribute`) together.

What do you think?

Reply:
Thanks for the response!

I would not merge `DefinedProctimeAttribute` and `DefinedProctimeAttribute` into a single interface. Both are for different types of timestamps (processing time and event time):

The idea is to have two interfaces:
- `DefinedProctimeAttribute` adds a timestamp attribute for processing time operations. The interface specifies the name of the new attribute. The source will simply add a virtual attribute, that fetches the current time when it is accessed.
- `DefinedRowtimeAttribute` (or its successor) specifies an existing attribute (via its name) to be an event time attribute and provides a watermark strategy (ascending, bounded-ooo, custom). If a `TableSource` specifies the interface, the scan will generate watermarks based on the existing field and the watermark strategy. During registration, we check that the field exists and has the right type. So instead of adding a field, we simply ensure that we have watermarks for it and change its type to be a time indicator field.

In this design, we `DefinedRowtimeAttribute` specifies the field and the watermark strategy as you proposed.

Reply:
Sorry for not stating it more clear. I was proposing having a tighter connection between `TimestampAssigner` and `DefinedProctimeAttribute`/ `DefinedRowtimeAttribute`, but not merging `DefinedProctimeAttribute`/ `DefinedRowtimeAttribute`.

In our use cases it is not enough to just have a trivial implementation of `TimeStampAssigner` to extract rowtime. We need to convert the corresponding field from `double` to `bigint`. It would be great if we can do it a little bit cleaner.

Reply:
Unfortunately we do need customer logics -- other people might need them too to take care issues like time zone differences.

Reply:
I assume you also need access to the full row, correct?

Reply:
As of today the answer is no -- we only need a single, top-level field. There might be use cases that have the rowtime in the nested fields. `TimestampAssigner` does come handy as the flexibility is quite important to us.

Maybe it makes sense to provide some default implementation of `TimestampAssigner` to extract a field in the row in order to make it easier to use?

Reply:
I see. Let's collect the requirements for the watermark and timestamp generation in the respective JIRA issue: https://issues.apache.org/jira/browse/FLINK-7548.